dd-sdk-android/src/main/cpp/native-lib.cpp

```diff
@@ -0,0 +1,194 @@
+                return abort_message;
+            } else {
+                // Not a cpp exception, assume a custom crash and act like C
+            }
+        }
+        catch (std::bad_cast& bc) {
+            // Not a cpp exception, assume a custom crash and act like C
+        }
+    }
+    // Assume C crash and print signal no and code
+    sprintf(abort_message, "Terminating with a C crash %d : %d", signo, siginfo->si_code);
+    return abort_message;
+}
+
+/// Main signal handling function.
+static void nativeCrashSignalHandler(int signo, siginfo* siginfo, void* ctxvoid) {
+    // Restoring an old handler to make built-in Android crash mechanism work.
+    sigaction(signo, &crashInContext->old_handlers[signo], nullptr);
+    // Log crash message
+    __android_log_print(ANDROID_LOG_ERROR, "NDK Playground", "%s", createCrashMessage(signo, siginfo));
+    // In some cases we need to re-send a signal to run standard bionic handler.
+    if (siginfo->si_code <= 0 || signo == SIGABRT) {
+        if (syscall(__NR_tgkill, getpid(), gettid(), signo) < 0) {
+            _exit(1);
+        }
+    }
+}
+
+/// like TestFairy.begin() but for crashes
+static void initializeNativeCrashHandler() {
+    // Check if already initialized
+    if (crashInContext) {
+        __android_log_print(ANDROID_LOG_INFO, "NDK Playground", "%s", "Native crash handler is already initialized.");
+        return;
+    }
+    // Initialize singleton crash handler context
+    crashInContext = static_cast<CrashInContext *>(malloc(sizeof(CrashInContext)));
+    memset(crashInContext, 0, sizeof(CrashInContext));
+    // Trying to register signal handler.
+    if (!registerSignalHandler(&nativeCrashSignalHandler, crashInContext->old_handlers)) {
+        deinitializeNativeCrashHandler();
+        __android_log_print(ANDROID_LOG_ERROR, "NDK Playground", "%s", "Native crash handler initialization failed.");
+        return;
+    }
+    __android_log_print(ANDROID_LOG_ERROR, "NDK Playground", "%s", "Native crash handler successfully initialized.");
+}
+
+/// Jni bindings
+extern "C" JNIEXPORT void JNICALL
+Java_com_datadog_android_core_internal_CoreFeature_initSignalHandler(
+        JNIEnv* env,
+        jobject /* this */) {
+    initializeNativeCrashHandler();
+}
+extern "C" JNIEXPORT void JNICALL
+Java_com_datadog_android_core_internal_CoreFeature_deinitSignalHandler(
+        JNIEnv* env,
+        jobject /* this */) {
+    deinitializeNativeCrashHandler();
+}
+/// Our custom test exception. Anything "publicly" inheriting std::exception will work
+class MyException : public std::exception {
+public:
+    const char* what() const noexcept override {
+        return "This is a really important crash message!";
+    }
+};
+extern "C" JNIEXPORT void JNICALL
+Java_com_datadog_android_sample_NavActivity_crashAndGetExceptionMessage(
+        JNIEnv* env,
+        jobject /* this */) {
+    throw MyException(); // This can be replaced with any foreign function call that throws.
+}
```

Comment:
So the NDK Crash reporter re-throws the exception to JVM? Is the JVM process kept alive in NDK-crash circumstances 😮?

On iOS there's not much to do while the process is crashing, so iOS crash reporters just write the report to file and we read it on next launch.

Reply:
Nope actually what I do here is just force an Exception in the Native part. Then that CrashHandler in the C++ file is catching this automatically, extracts the information and then forwards it to the JVM logger through this call : `__android_log_print(...`
Now I am trying to find a way to actually re - throw that exception in JVM (and this way it will be caught by our SDK CrashHandler) or find a way to actually call a method on our SDK from Native and propagate that.

Reply:
Then I'm not sure if I get it right - the `crashAndGetExceptionMessage()` simulates throwing Java Exception from C++ to JVM? If so, then I'm not sure if this is what we need. IMO the goal is to catch crashes coming from those signals:
```cpp
/// Caught signals
 static const int SIGNALS_TO_CATCH[] = {
         SIGABRT,
         SIGBUS,
         SIGFPE,
         SIGSEGV,
         SIGILL,
         SIGSTKFLT,
         SIGTRAP,
 };
```
And to simulate some of them, we need to execute bad native code. For instance, the `SIGTRAP` which I POCed on iOS can be issued due to [array out of bound access](https://developer.apple.com/documentation/code_diagnostics/undefined_behavior_sanitizer/out-of-bounds_array_access), for example this `C` code:
```c
int array[5];
for (int i = 0; i <= 5; ++i) {
    array[i] += 1; // Error: out-of-bounds access on the last iteration
}
```
And if Android NDK follows the same process architecture as iOS, there won't be any chance to re-throw an exception to JVM, as the entire process is being corrupted. But I don't know the NDK, I might be totally wrong 🙌.